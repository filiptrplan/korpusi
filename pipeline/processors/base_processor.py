class BaseProcessor:
    """
    Base class for all processors. 
    All processors should inherit from this class and implement the process method.
    """
    def __init__(self, song: any, name: str, mapping = None):
        """    
        any song: The song to process. 
        str name: The name of the processor. This is the name of the field that the results will be stored in.
        """
        self.song = song
        self.name = name
        self.mapping = mapping

    def process(self):
        """The main function of the processor. It should spit out the results in dictionary format."""
        raise NotImplementedError

    def get_name(self):
        """Returns the name of the processor."""
        return self.name
    
    def get_mapping(self):
        """Returns the mapping of the processor."""
        if self.mapping is None:
            raise ValueError("Mapping not set")
        return self.mapping