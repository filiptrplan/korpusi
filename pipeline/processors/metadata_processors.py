import csv
import os.path


class CSVMetadataProcessor:
    csv_content = []
    csv_header = []

    def __init__(self, csv_file_path: str):
        """
        @param csv_file_path: This should be the path of the CSV file containing the metadata. The first row should be the
        header specifying different columns. This class will match the `filename` column and then output the other columns
        as metadata. Call the `process` function to get the metadata for a specific file.
        """
        separator = ","
        with open(csv_file_path, "r", encoding="utf-8") as f:
            first_line = f.readline()
            if "," in first_line:
                separator = ","
            elif ";" in first_line:
                separator = ";"
        with open(csv_file_path, "r", encoding="utf-8") as f:
            reader = csv.DictReader(f, delimiter=separator)
            for row in reader:
                self.csv_content.append(row)
            self.csv_header = reader.fieldnames
            if "filename" not in self.csv_header:
                raise ValueError("CSV file must have a column named 'filename'")

    def get_mapping(self):
        properties_dict = {
            "filename": {"enabled": False},
        }

        for column in self.csv_header:
            if column != "filename":
                properties_dict[column] = {
                    "type": "keyword",
                    "fields": {"text": {"type": "text"}},
                }

        return {"properties": properties_dict}

    def process(self, file_path: str):
        """
        Returns the metadata for the file.
        """
        filename = os.path.basename(file_path)
        # Get only the file name without the extension
        filename = os.path.splitext(filename)[0]
        for row in self.csv_content:
            # Get the row filename without the extension
            row_filename = os.path.splitext(row["filename"])[0]
            if row_filename == filename:
                return row
        return {}
