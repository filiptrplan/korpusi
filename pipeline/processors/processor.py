import music21

class Processor:
    """
    Base class for all processors. 
    All processors should inherit from this class and implement the process method.
    """
    def __init__(self, song: music21.stream.Stream, name: str):
        """    
        music21.stream.Stream song: The song to process. This should be output from music21.converter.parse.

        str name: The name of the processor. This is the name of the field that the results will be stored in.
        """
        self.song = song
        self.name = name


    def process(self):
        """The main function of the processor. It should spit out the results in JSON."""
        raise NotImplementedError
    
    
    def get_results(self):
        """Returns the results of the process function in JSON."""
        raise NotImplementedError


    def get_name(self):
        """Returns the name of the processor."""
        return self.name
