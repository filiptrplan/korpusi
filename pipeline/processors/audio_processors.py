import os

import soundfile

from helpers import check_audio_extension_allowed
from processors.base_processor import BaseProcessor


class AudioProcessor(BaseProcessor):
    """
    Base class for all audio processors.
    All processors that process audio files should inherit from this class and implement the process method.
    """

    def __init__(
        self,
        song: str,
        algorithm_name: str = None,
        feature_name: str = None,
        mapping=None,
    ):
        """
        any song: The path to the song to process.
        str name: The name of the processor. This is the name of the field that the results will be stored in.
        """
        # Song is None when using it for mapping generation
        if song is not None:
            if not isinstance(song, str):
                raise ValueError("Song path must be a string")
            if not os.path.exists(song):
                raise ValueError("Song path does not exist")
            if not check_audio_extension_allowed(song):
                raise ValueError("Song must be an audio file")
        if feature_name is None:
            feature_name = self.__class__.__name__
        if algorithm_name is None:
            raise ValueError("Algorithm name must be specified.")
        self.algorithm_name = algorithm_name
        super().__init__(song, feature_name, mapping)

    def process(self):
        """The main function of the processor. It should spit out the results in dictionary format or a single value."""
        raise NotImplementedError("Subclasses must implement this method")

    def get_algorithm_name(self):
        """Returns the name of the algorithm used for the processor."""
        return self.algorithm_name


class AudioFileInfoProcessor(AudioProcessor):
    """Gets the file information of the song like the duration, sample rate, and bit rate."""

    def __init__(self, song: any):
        super().__init__(song, "file_info", "sample_rate")
        self.mapping = {
            "properties": {
                "sample_rate": {"type": "float"},
                "duration": {"type": "float"},
                "encoding_subtype": {"type": "keyword"},
            }
        }

    def process(self):
        file = soundfile.SoundFile(self.song)
        return {
            "sample_rate": file.samplerate,
            "duration": file.frames / file.samplerate,
            "encoding_subtype": file.subtype,
        }


class AudioBPMProcessor(AudioProcessor):
    """Gets the BPM of the song."""

    def __init__(self, song: any):
        super().__init__(song, "essentia_multifeature", "bpm")
        self.mapping = {
            "properties": {"bpm": {"type": "float"}, "beat_ticks": {"type": "float"}}
        }

    def process(self):
        import essentia.standard

        # let the loader resample here! if we include the original sample rate it ruins the accuracy
        # of the algorithm
        loader = essentia.standard.MonoLoader(filename=self.song)
        audio = loader()

        rhythm_extractor = essentia.standard.RhythmExtractor2013(method="multifeature")
        beats = rhythm_extractor(audio)

        bpm = round_floats(beats[0])
        beat_ticks = round_floats(beats[1].tolist())

        return {"bpm": bpm, "beat_ticks": beat_ticks}


class AudioPitchContourProcessor(AudioProcessor):
    """Gets the pitch contour of the song."""

    def __init__(self, song: any):
        super().__init__(song, "pesto", "pitch_contour")
        self.mapping = {
            "properties": {
                "pitch_contour_hz_voice": {"type": "float"},
                "pitch_contour_hz_instrumental": {"type": "float"},
                "time_step_ms": {"type": "float"},
            }
        }

    def process(self):
        import torchaudio
        import torch
        from pesto import predict

        file_extension = self.song.split(".")[-1]
        rest_of_path = self.song[: -len(file_extension) - 1]
        step_size = 10.0

        voice_path = rest_of_path + ".vocals.mp3"
        instrumental_path = rest_of_path + ".accompaniment.mp3"
        if not os.path.exists(voice_path):
            raise ValueError(
                f"{voice_path} does not exist. Please run the voice extraction first. Refer to extract_voice.md for more information."
            )

        if not os.path.exists(instrumental_path):
            raise ValueError(
                f"{instrumental_path} does not exist. Please run the voice extraction first. Refer to extract_voice.md for more information."
            )

        device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")

        x, sr = torchaudio.load(voice_path)
        x = x.mean(dim=0)  # PESTO takes mono audio as input
        x = x.to(device)

        timesteps, predictions_voice, confidence, activations = predict(
            x, sr, step_size
        )

        x, sr = torchaudio.load(instrumental_path)
        x = x.mean(dim=0)
        x = x.to(device)
        timesteps, predictions_instrumental, confidence, activations = predict(
            x, sr, step_size
        )

        return {
            "pitch_contour_hz_voice": round_floats(predictions_voice.tolist()),
            "pitch_contour_hz_instrumental": round_floats(
                predictions_instrumental.tolist()
            ),
            "time_step_ms": step_size,
        }


class AudioChordProcessor(AudioProcessor):
    """Gets the chord progression of the song."""

    def __init__(self, song: any):
        super().__init__(song, "autochord", "chords")
        self.mapping = {
            "properties": {
                "chord_name": {"type": "keyword"},
                "chord_start": {"type": "float"},
                "chord_end": {"type": "float"},
            }
        }

    def process(self):
        import autochord

        output = autochord.recognize(self.song)
        chord_names = [x[2] for x in output]
        chord_starts = round_floats([x[0] for x in output])
        chord_ends = round_floats([x[1] for x in output])
        return {
            "chord_name": chord_names,
            "chord_start": chord_starts,
            "chord_end": chord_ends,
        }


class AudioRMSProcessor(AudioProcessor):
    def __init__(self, song: any):
        super().__init__(song, "rms", "loudness")
        self.mapping = {
            "properties": {
                "loudness_total": {"type": "float"},
                "loudness_vocals": {"type": "float"},
                "loudness_instrumental": {"type": "float"},
                "timestep_seconds": {"type": "float"},
            }
        }

    def rms(self, song):
        import essentia.standard as es

        sample_rate = get_sample_rate(song)
        frame_size = int(sample_rate / 16)
        hop_size = int(frame_size / 2)

        loader = es.MonoLoader(filename=song, sampleRate=sample_rate)
        audio = loader()

        rms = es.RMS()
        rms_values = []

        for frame in es.FrameGenerator(audio, frameSize=frame_size, hopSize=hop_size):
            rms_values.append(rms(frame))

        rms_timestep_seconds = hop_size / sample_rate

        return (rms_values, rms_timestep_seconds)

    def process(self):
        file_extension = self.song.split(".")[-1]
        rest_of_path = self.song[: -len(file_extension) - 1]

        voice_path = rest_of_path + ".vocals.mp3"
        instrumental_path = rest_of_path + ".accompaniment.mp3"

        (rms_values_total, timestep) = self.rms(self.song)
        (rms_values_vocals, _) = self.rms(voice_path)
        (rms_values_instrumental, _) = self.rms(instrumental_path)

        return {
            "loudness_total": rms_values_total,
            "loudness_vocals": rms_values_vocals,
            "loudness_instrumental": rms_values_instrumental,
            "timestep_seconds": timestep,
        }


class AudioKeyExtractProcessor(AudioProcessor):
    def __init__(self, song: any):
        super().__init__(song, "essentia_key_extractor", "key")
        self.mapping = {
            "properties": {
                "key": {"type": "keyword"},
                "scale": {"type": "keyword"},
                "confidence": {"type": "float"},
            }
        }

    def process(self):
        import essentia.standard as es

        loader = es.MonoLoader(filename=self.song)
        audio = loader()
        key_extract = es.KeyExtractor()

        (key, scale, confidence) = key_extract(audio)

        return {"key": key, "scale": scale, "confidence": confidence}


def round_floats(o):
    if isinstance(o, float):
        return round(o, 2)
    if isinstance(o, dict):
        return {k: round_floats(v) for k, v in o.items()}
    if isinstance(o, (list, tuple)):
        return [round_floats(x) for x in o]
    return o


def get_sample_rate(song):
    import soundfile

    file = soundfile.SoundFile(song)
    return file.samplerate
