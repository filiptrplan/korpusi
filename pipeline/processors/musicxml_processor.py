import music21
from processors.base_processor import BaseProcessor


class MusicXMLProcessor(BaseProcessor):
    """
    Base class for all musicXML processors.
    All processors that process musicXML files should inherit from this class and implement the process method.
    """

    def __init__(self, song: music21.stream.Stream, name: str = None, mapping=None):
        """
        music21.stream.Stream song: The song to process. This should be output from music21.converter.parse.

        The name of the processor is automatically set to the name of the class. It should be hard-coded in the class to
        something meaningful to better identify the results.
        """
        if name is None:
            name = self.__class__.__name__
        super().__init__(song, name, mapping)

    def process(self):
        """The main function of the processor. It should spit out the results in dictionary format or a single value."""
        raise NotImplementedError

    def get_name(self):
        """Returns the name of the processor."""
        return self.name
