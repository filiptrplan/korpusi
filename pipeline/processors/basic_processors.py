from music21 import stream, tempo, note, interval, metadata, key
from processors.musicxml_processor import MusicXMLProcessor


class KeyProcessor(MusicXMLProcessor):
    """Get the key signature of the song."""

    song: stream.Stream

    def __init__(self, song: stream.Stream, feature_name="key"):
        super().__init__(song, feature_name)
        self.mapping = {
            "properties": {
                "most_certain_key": {
                    "type": "keyword",
                    "fields": {"text": {"type": "text"}},
                },
                "alternate_keys": {
                    "type": "keyword",
                    "fields": {"text": {"type": "text"}},
                },
            }
        }

    def process(self) -> dict[str, list[str] | str]:
        song_key: key.Key = self.song.analyze("key")
        return {
            "most_certain_key": song_key.tonicPitchNameWithCase,
            "alternate_keys": [
                x.tonicPitchNameWithCase for x in song_key.alternateInterpretations[0:4]
            ],
            # get the first 4 alternate keys
        }


class TimeSignatureProcessor(MusicXMLProcessor):
    """Get all time signatures in the song."""

    song: stream.Stream

    def __init__(self, song: stream.Stream, feature_name="time_signature"):
        super().__init__(song, feature_name)
        self.mapping = {"type": "keyword", "fields": {"text": {"type": "text"}}}

    def process(self) -> list[str]:
        time_signatures = set()

        # Iterate through all parts and measures to find time signatures
        for part in self.song.parts:
            measures = part.getElementsByClass(stream.Measure)
            for measure in measures:
                if measure.timeSignature is not None:
                    time_signatures.add(str(measure.timeSignature.ratioString))

        # Convert set to sorted list for consistent ordering
        return sorted(list(time_signatures))


class TempoProcessor(MusicXMLProcessor):
    """Get the tempo of the song."""

    song: stream.Stream

    def __init__(self, song: stream.Stream, feature_name="tempo"):
        super().__init__(song, feature_name)
        self.mapping = {"type": "long"}

    def process(self) -> str:
        tempo_marks = []
        for sub_stream in self.song.recurse(streamsOnly=True, includeSelf=True):
            found = sub_stream.getElementsByClass(tempo.MetronomeMark)
            for x in found:
                tempo_marks.append(x)

        if len(tempo_marks) == 0:
            tempo_str = None
        elif tempo_marks[0].number is not None:
            tempo_str = str(tempo_marks[0].number)
        else:
            tempo_str = None
        return tempo_str


class AmbitusProcessor(MusicXMLProcessor):
    """Gets the 'ambitus' of a song. This is the range of the song."""

    song: stream.Stream

    def __init__(self, song: stream.Stream, feature_name="ambitus"):
        super().__init__(song, feature_name)
        self.mapping = {
            "properties": {
                "min_note": {
                    "type": "long",
                },
                "max_note": {
                    "type": "long",
                },
                "ambitus_semitones": {"type": "long"},
            }
        }

    def process(self):
        notes = self.song.flatten().getElementsByClass(note.Note)
        min_note = note.Note(
            "C8"
        )  # je tole v redu? Ponavadi gre od 0 do 8 po oktavah navzgor
        max_note = note.Note("A0")
        for n in notes:
            max_note = max(n, max_note)
            min_note = min(n, min_note)

        my_interval = interval.Interval(noteStart=min_note, noteEnd=max_note)

        result = {
            "min_note": min_note.pitch.midi,
            "max_note": max_note.pitch.midi,
            "ambitus_semitones": my_interval.semitones,
        }

        return result


class MetadataProcessor(MusicXMLProcessor):
    """Gets the metadata of a song."""

    song: stream.Stream

    def __init__(self, song: stream.Stream, feature_name="metadata"):
        super().__init__(song, feature_name)
        self.mapping = {
            "type": "object",
        }

    def process(self):
        my_metadata = self.song.getElementsByClass(metadata.Metadata)[0]
        my_metadata_tuples = my_metadata.all()
        my_metadata_dict = {}
        for t in my_metadata_tuples:
            my_metadata_dict[t[0]] = t[1]
        my_metadata_dict.pop("filePath")
        my_metadata_dict.pop("fileFormat")
        my_metadata_dict.pop("software")
        return my_metadata_dict


class DurationProcessor(MusicXMLProcessor):
    """Gets the duration of a song."""

    song: stream.Stream

    def __init__(self, song: stream.Stream, feature_name="duration"):
        super().__init__(song, feature_name)
        self.mapping = {
            "properties": {
                "measures": {"type": "long"},
                "beats": {"type": "long"},
            }
        }

    def process(self):
        beats = self.song.duration.quarterLength
        measures = 0
        for x in self.song.flatten():
            if x.measureNumber is not None and x.measureNumber > measures:
                measures = x.measureNumber
        return {"measures": measures, "beats": beats}
