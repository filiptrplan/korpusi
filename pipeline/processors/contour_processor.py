import music21

from processors.musicxml_processor import MusicXMLProcessor

class ContourProcessor(MusicXMLProcessor):
    """Gets the contour of the song. It describes the shape of the melody(when notes ascend or descend)."""

    song: music21.stream.Stream
    def __init__(self, song: music21.stream.Stream, name='contour'):
        super().__init__(song, name)
        self.mapping = {
            'properties': {
                'melodic_contour_semi': {
                    'type': 'long'
                },
                'melodic_contour_string': {
                    'type': 'text'
                },
                'measure_starts': {
                    'type': 'long'
                }
            }
        }

    def process(self):
        pitch_values = []
        measure_numbers = [] # this is done because you can't set a measure number to a note and a note in a chord doesn't have one
        beats = []
        for x in self.song.parts[0].flatten():
            if len(measure_numbers) != 0:
                if measure_numbers[len(measure_numbers) - 1] == x.measureNumber and beats[len(beats) - 1] == x.beat:
                    continue # prevents adding two notes that are played at the same time
            if isinstance(x, music21.note.Note):
                pitch_values.append(x)
                measure_numbers.append(x.measureNumber)
                beats.append(x.beat)
            elif isinstance(x, music21.chord.Chord):
                note = x.sortAscending()[len(x) - 1]
                measure_numbers.append(x.measureNumber)
                pitch_values.append(note)
                beats.append(x.beat)
        melodic_contour = []
        melodic_contour_string = ""
        for i in range(len(pitch_values) - 1):
            pitch_interval = music21.interval.Interval(pitch_values[i], pitch_values[i+1])
            semi = pitch_interval.semitones
            melodic_contour.append(semi)
            
        melodic_contour_string_absolute = ' '.join([x.nameWithOctave for x in pitch_values])
            
        melodic_contour_string = ""
        for x in melodic_contour:
            if x > 0: 
                melodic_contour_string += "U"
            elif x < 0: 
                melodic_contour_string += "D"
            else: 
                melodic_contour_string += "S"
            melodic_contour_string += " " # this is here so we can better search for the contour

        measure_starts = [i for i in range(len(pitch_values)) if measure_numbers[i] != measure_numbers[i-1]]

        return {
            'melodic_contour_semi': melodic_contour,
            'melodic_contour_string_relative': melodic_contour_string,
            'melodic_contour_string_absolute': melodic_contour_string_absolute,
            'measure_starts': measure_starts
        }
    