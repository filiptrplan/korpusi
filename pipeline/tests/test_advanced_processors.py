import os

import music21
from processors.contour_processor import (
    ContourProcessor,
    RhythmProcessor,
    NGramRhythmProcessor,
    NGramPitchProcessor,
)


def song():
    """
    Returns a music21 stream object of the test song.
    """
    base_path = os.path.dirname(__file__)
    path = os.path.join(base_path, "test.musicxml")

    return music21.converter.parse(path)


class TestAdvancedProcessors:
    """Tests processors from contour_processor.py"""

    def test_contour_processor(self, snapshot):
        contour_processor = ContourProcessor(song())
        result = contour_processor.process()
        assert result == snapshot

    def test_rhythm_processor(self, snapshot):
        rhythm_processor = RhythmProcessor(song())
        result = rhythm_processor.process()
        assert result == snapshot

    def test_rhythm_ngram_processor(self, snapshot):
        rhythm_ngram_processor = NGramRhythmProcessor(song())
        result = rhythm_ngram_processor.process()
        assert result == snapshot

    def test_pitch_ngram_processor(self, snapshot):
        pitch_ngram_processor = NGramPitchProcessor(song())
        result = pitch_ngram_processor.process()
        assert result == snapshot
