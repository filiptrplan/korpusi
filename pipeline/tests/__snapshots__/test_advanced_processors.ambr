# serializer version: 1
# name: TestAdvancedProcessors.test_contour_processor
  dict({
    'measure_starts': list([
      0,
      2,
      4,
      6,
      8,
      10,
      12,
      14,
      16,
      18,
      20,
      22,
      24,
      27,
      29,
      32,
      34,
      36,
      38,
      40,
      42,
      44,
      46,
      48,
      50,
      52,
      54,
      56,
      58,
      60,
      62,
      65,
      66,
    ]),
    'melodic_contour_string': 'D U D U S S D S D U D U S S D U D U D U S S D S S D U D U U D U D U U S D U D S S D U S D U D S S D U S D U D S S D U S D S U D U D ',
    'melodic_contour_string_absolute': '67 64 67 64 67 67 67 65 65 62 65 62 65 65 65 64 67 64 67 64 67 67 67 65 65 65 62 65 62 65 67 62 65 64 67 72 72 67 71 69 69 69 65 71 71 65 69 67 67 67 64 69 69 64 67 65 65 65 62 67 67 65 65 67 62 65 64',
    'melodic_contour_string_relative': '-3 3 -3 3 0 0 -2 0 -3 3 -3 3 0 0 -1 3 -3 3 -3 3 0 0 -2 0 0 -3 3 -3 3 2 -5 3 -1 3 5 0 -5 4 -2 0 0 -4 6 0 -6 4 -2 0 0 -3 5 0 -5 3 -2 0 0 -3 5 0 -2 0 2 -5 3 -1',
  })
# ---
# name: TestAdvancedProcessors.test_pitch_ngram_processor
  dict({
    'frequency_histogram': dict({
      'M+2 P-4 m+3': 2,
      'M+2 P-4 m+3 M-2': 2,
      'M+3 M-3 M+3': 3,
      'M+3 M-3 M+3 P+1': 2,
      'M+3 M-3 M+3 P+1 P+1': 2,
      'M+3 M-3 M+3 P+1 P+1 M-2': 2,
      'M+3 M-3 M+3 P+1 P+1 M-2 P+1': 2,
      'M+3 P+1 P+1': 2,
      'M+3 P+1 P+1 M-2': 2,
      'M+3 P+1 P+1 M-2 P+1': 2,
      'M-2 P+1 P+1': 3,
      'M-2 P+1 P+1 m-3': 3,
      'M-2 P+1 P+1 m-3 P+4': 2,
      'M-2 P+1 P+1 m-3 P+4 P+1': 2,
      'M-3 M+3 M-3': 2,
      'M-3 M+3 M-3 M+3': 2,
      'M-3 M+3 M-3 M+3 P+1': 2,
      'M-3 M+3 M-3 M+3 P+1 P+1': 2,
      'M-3 M+3 M-3 M+3 P+1 P+1 M-2': 2,
      'M-3 M+3 M-3 M+3 P+1 P+1 M-2 P+1': 2,
      'M-3 M+3 P+1': 2,
      'M-3 M+3 P+1 P+1': 2,
      'M-3 M+3 P+1 P+1 M-2': 2,
      'M-3 M+3 P+1 P+1 M-2 P+1': 2,
      'P+1 M-2 P+1': 3,
      'P+1 P+1 M-2': 3,
      'P+1 P+1 M-2 P+1': 2,
      'P+1 P+1 m-3': 3,
      'P+1 P+1 m-3 P+4': 2,
      'P+1 P+1 m-3 P+4 P+1': 2,
      'P+1 P-4 m+3': 2,
      'P+1 m-3 P+4': 2,
      'P+1 m-3 P+4 P+1': 2,
      'P+1 m-3 m+3': 2,
      'P+1 m-3 m+3 m-3': 2,
      'P+1 m-3 m+3 m-3 m+3': 2,
      'P+4 P+1 P-4': 3,
      'P+4 P+1 P-4 m+3': 2,
      'P-4 m+3 M-2': 3,
      'm+3 m-3 m+3': 2,
      'm-3 P+4 P+1': 2,
      'm-3 m+3 m-3': 2,
      'm-3 m+3 m-3 m+3': 2,
    }),
  })
# ---
# name: TestAdvancedProcessors.test_rhythm_ngram_processor
  dict({
    'frequency_histogram': dict({
      '1/1 1/2 1/1': 25,
      '1/1 1/2 1/1 1/2': 25,
      '1/1 1/2 1/1 1/2 1/1': 23,
      '1/1 1/2 1/1 1/2 1/1 1/2': 23,
      '1/1 1/2 1/1 1/2 1/1 1/2 1/1': 21,
      '1/1 1/2 1/1 1/2 1/1 1/2 1/1 1/2': 21,
      '1/1 1/2 1/1 1/2 1/1 1/2 1/1 1/2 1/1': 19,
      '1/1 1/2 1/1 1/2 1/1 1/2 1/1 1/2 1/2': 2,
      '1/1 1/2 1/1 1/2 1/1 1/2 1/2': 2,
      '1/1 1/2 1/1 1/2 1/1 1/2 1/2 1/2': 2,
      '1/1 1/2 1/1 1/2 1/1 1/2 1/2 1/2 1/2': 2,
      '1/1 1/2 1/1 1/2 1/2': 2,
      '1/1 1/2 1/1 1/2 1/2 1/2': 2,
      '1/1 1/2 1/1 1/2 1/2 1/2 1/2': 2,
      '1/1 1/2 1/2': 3,
      '1/1 1/2 1/2 1/2': 3,
      '1/1 1/2 1/2 1/2 1/2': 3,
      '1/1 1/2 1/2 1/2 1/2 1/1': 2,
      '1/1 1/2 1/2 1/2 1/2 1/1 1/2': 2,
      '1/2 1/1 1/2': 27,
      '1/2 1/1 1/2 1/1': 24,
      '1/2 1/1 1/2 1/1 1/2': 24,
      '1/2 1/1 1/2 1/1 1/2 1/1': 22,
      '1/2 1/1 1/2 1/1 1/2 1/1 1/2': 22,
      '1/2 1/1 1/2 1/1 1/2 1/1 1/2 1/1': 20,
      '1/2 1/1 1/2 1/1 1/2 1/1 1/2 1/1 1/2': 20,
      '1/2 1/1 1/2 1/1 1/2 1/1 1/2 1/2': 2,
      '1/2 1/1 1/2 1/1 1/2 1/1 1/2 1/2 1/2': 2,
      '1/2 1/1 1/2 1/1 1/2 1/2': 2,
      '1/2 1/1 1/2 1/1 1/2 1/2 1/2': 2,
      '1/2 1/1 1/2 1/1 1/2 1/2 1/2 1/2': 2,
      '1/2 1/1 1/2 1/2': 3,
      '1/2 1/1 1/2 1/2 1/2': 3,
      '1/2 1/1 1/2 1/2 1/2 1/2': 3,
      '1/2 1/1 1/2 1/2 1/2 1/2 1/1': 2,
      '1/2 1/1 1/2 1/2 1/2 1/2 1/1 1/2': 2,
      '1/2 1/2 1/1': 2,
      '1/2 1/2 1/1 1/2': 2,
      '1/2 1/2 1/2': 6,
      '1/2 1/2 1/2 1/1': 2,
      '1/2 1/2 1/2 1/1 1/2': 2,
      '1/2 1/2 1/2 1/2': 3,
      '1/2 1/2 1/2 1/2 1/1': 2,
      '1/2 1/2 1/2 1/2 1/1 1/2': 2,
    }),
  })
# ---
# name: TestAdvancedProcessors.test_rhythm_processor
  dict({
    'measure_starts': list([
      0,
      2,
      4,
      6,
      8,
      10,
      12,
      14,
      16,
      18,
      20,
      22,
      24,
      27,
      29,
      32,
      34,
      36,
      38,
      40,
      42,
      44,
      46,
      48,
      50,
      52,
      54,
      56,
      58,
      60,
      62,
      65,
      66,
    ]),
    'rhythm_string': '1/1 1/2 1/1 1/2 1/1 1/2 1/1 1/2 1/1 1/2 1/1 1/2 1/1 1/2 1/1 1/2 1/1 1/2 1/1 1/2 1/1 1/2 1/1 1/2 1/2 1/2 1/2 1/1 1/2 1/2 1/2 1/2 1/1 1/2 1/1 1/2 1/1 1/2 1/1 1/2 1/1 1/2 1/1 1/2 1/1 1/2 1/1 1/2 1/1 1/2 1/1 1/2 1/1 1/2 1/1 1/2 1/1 1/2 1/1 1/2 1/1 1/2 1/2 1/2 1/2 3/2 3/2',
  })
# ---
